Minimize memory footprint whereever possible, add joblib support for …

dPys opened the pull request:
@arokem
 
 [PR #53](https://github.com/nipreps/eddymotion/issues/53) and [PR #60](https://github.com/nipreps/eddymotion/pull/60)

Substantial speedup and less virtual memory overhead by avoiding/deleting unneeded objects + garbage collection

pep8speaks commented:
Hello @dPys, Thank you for submitting the Pull Request !

* In the file [`dipy/reconst/sfm.py`](https://github.com/dipy/dipy/blob/6578757359ec3654ef4f69cdaba679894367dc97/dipy/reconst/sfm.py):

> [Line 498:60](https://github.com/dipy/dipy/blob/6578757359ec3654ef4f69cdaba679894367dc97/dipy/reconst/sfm.py#L498): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent



Do see the [DIPY coding Style guideline](https://github.com/dipy/dipy/blob/master/doc/devel/coding_style_guideline.rst)

arokem commented on dipy/reconst/sfm.py:
```diff
@@ -458,21 +475,45 @@ def fit(self, data, mask=None):
-                                                       fit_it).coef_
+        if n_threads > 1:
+            from collections import OrderedDict
+            try:
```
We should use https://github.com/dipy/dipy/blob/master/dipy/utils/optpkg.py#L18 instead of this construction. 

Check out the sklearn import at the top of this file.

dPys replied:
Isn't the Collections module is a built-in module that comes with a base installation of Python? For now I've just moved the import up to the top of the file.

arokem replied:
For the joblib import

arokem commented on dipy/reconst/sfm.py:
```diff
@@ -458,21 +475,45 @@ def fit(self, data, mask=None):
+                          backend='multiprocessing') as parallel:
+                out = parallel(
+                    delayed(self.fit_solver2voxels)(isopredict,
+                                                           vox_data, vox,
```
deindent for match with line before: 
```suggestion
                                                    vox_data, vox,
```

arokem commented on dipy/reconst/sfm.py:
```diff
@@ -458,21 +475,45 @@ def fit(self, data, mask=None):
+                out = parallel(
+                    delayed(self.fit_solver2voxels)(isopredict,
+                                                           vox_data, vox,
+                                                    True) for
```
here as well

dPys replied:
For joblib, I went ahead and used the https://github.com/dipy/dipy/blob/master/dipy/utils/optpkg.py#L18 construction as suggested.

arokem commented:
LGTM! But we do also need to add installation of joblib here:

https://github.com/dipy/dipy/blob/master/dipy/info.py#L109

I think that it's pretty uncontroversial to take it on as a dependency (pure Python, very well-supported), but curious to hear if others have any objections.

arokem reviewed (commented):
Some additional review comments

skoudoro commented on dipy/reconst/sfm.py:
```diff
@@ -420,7 +420,24 @@ def design_matrix(self):
+            with warnings.catch_warnings():
+                warnings.simplefilter("ignore")
+
+        if parallel is True:
```
I recommend avoiding using `is`.  It is confusing for many people. for example
`if 1 is True` returns `False` when many people expect `True` due to a misunderstanding of `is` .

You can do `if parallel:` instead

dPys replied:
Sounds good to me, assuming that `parallel` will always be set to either `True` or `False` and not `None` or some other number/character(s) (e.g. `5`) ...  I've been in the (bad) habit of using `is` recently to avoid the latter scenario, which can cause issues in nipype. I bet we can at least both agree against `parallel == True` haha

skoudoro replied:
`parallel == True` not really pythonic 😄 ... if the type is important, you can use `isinstance`

`if parallel and isinstance(parallel, bool):`

skoudoro commented on dipy/reconst/sfm.py:
```diff
@@ -420,7 +420,24 @@ def design_matrix(self):
+        # parameters at all-zeros, and avoid nasty sklearn errors:
+        if not (np.any(~np.isfinite(vox_data)) or np.all(
+                vox_data == 0)):
+            with warnings.catch_warnings():
+                warnings.simplefilter("ignore")
```
what is the goal of this?

dPys replied:
Was in the original implementation I think? My guess would be that it suppresses icky sklearn errors on the solver fit.

skoudoro commented on dipy/reconst/sfm.py:
```diff
@@ -420,7 +420,24 @@ def design_matrix(self):
+                self.design_matrix, vox_data - isopredict[vox]
+            ).coef_
+
+    def fit(self, data, mask=None, n_threads=1):
```
threads or processes?  our conventional names are `num_processes` or `num_threads`

dPys replied:
Ideally this would be `num_threads` then, thanks for the clarification. Ultimately we'll probably want to switch the joblib backend to loky/cloudpickle once joblib figures out how to play nice with pickle5 compat...

skoudoro replied:
I just see tht you are using `backend='multiprocessing'`  so it should be `num_processes`. The multiprocessing module does not use threads. It spawns multi processes.

skoudoro commented on dipy/reconst/sfm.py:
```diff
@@ -458,21 +475,45 @@ def fit(self, data, mask=None):
-                    warnings.simplefilter("ignore")
-                    flat_params[vox] = self.solver.fit(self.design_matrix,
-                                                       fit_it).coef_
+        if n_threads > 1:
```
I recommend using this before:

```python
from dipy.utils.multiproc import determine_num_processes

num_processes = determine_num_processes(num_processes)
if num_processes > 1:
    ....
```

more infos [here](https://github.com/dipy/dipy/blob/master/dipy/utils/multiproc.py#L7)

dPys replied:
Love it

skoudoro reviewed (commented):
Hi @dPys,

Thanks for this update! See below for some additional comments.

no objection concerning @arokem proposal to add `joblib` in our dependency tree.

Garyfallidis commented:
I would like to see numbers. What are the actual benefits of using joblib? How much acceleration do we get? When we tested a couple of years ago it was not performing well.

dPys commented:
> I would like to see numbers. What are the actual benefits of using joblib? How much acceleration do we get? When we tested a couple of years ago it was not performing well.

It's embarrassingly parallel so should scale per amdahl's law, etc. At least on the basis of runtime duration, here's what `test_sfm` clocks in with:
```
iso: <class 'dipy.reconst.sfm.ExponentialIsotropicModel'>
sffit1: 0.2s. n_proc: 1
sffit2: 9.1s. n_proc: 1
sffit3: 18.4s. n_proc: 1

iso: <class 'dipy.reconst.sfm.ExponentialIsotropicModel'>
sffit1: 0.4s. n_proc: 4
sffit2: 5.1s. n_proc: 4
sffit3: 9.8s. n_proc: 4

iso: <class 'dipy.reconst.sfm.ExponentialIsotropicModel'>
sffit1: 0.4s. n_proc: 8
sffit2: 3.6s. n_proc: 8
sffit3: 6.5s. n_proc: 8
```

I don't know much about how it performed years ago, but I like joblib in general for the following reasons:
1) Standard library only / no dependencies.
2) Well-tested.
3) Minimalism + easy to implement with existing code + requires only local imports.
4) Supports multiple parallel backends (i.e. no just multiproc)
5) Caching features are also easy to implement and can be a nice solution for responsibly offloading memory burden to I/O.

dPys commented:
Pushed changes @skoudoro and @arokem . Please take a glance when you get a chance!

skoudoro commented on dipy/reconst/sfm.py:
```diff
@@ -458,21 +478,42 @@ def fit(self, data, mask=None):
+        if not num_processes:
+            num_processes = determine_num_processes(num_processes)
+
+        if num_processes > 1:
```
`if num_processes > 1 and has_joblib:` instead

dPys replied:
Added

skoudoro commented on dipy/reconst/sfm.py:
```diff
@@ -18,21 +18,25 @@
 from dipy.reconst.base import ReconstModel, ReconstFit
 from dipy.reconst.cache import Cache
 from dipy.core.onetime import auto_attr
+from collections import OrderedDict
```
Can you put the import of `gc` and `collections` with `warnings`. We try to classiffy them.
1. standard library
2. third party library
3. local import

dPys replied:
Fixed!

skoudoro commented on dipy/reconst/tests/test_sfm.py:
```diff
@@ -30,27 +33,40 @@ def test_sfm():
-            np.zeros(sfmodel.design_matrix[0].shape[-1]))
+    for n_threads in [1, 2]:
+        for iso in [sfm.ExponentialIsotropicModel, None]:
+            print(f"\niso: {iso}")
```
The `logging` module is preferred to `print`. Can you confirm @arokem?

dPys replied:
On my most recent commit, I've gone ahead and just commented these `print` lines out for now since they probably dont need to be run every time.

arokem replied:
Yeah - in this case, I would remove the print statements altogether. In general, when users need feedback, we prefer to use logging.

dPys replied:
Cool, removed

skoudoro reviewed (commented):
Thanks for the update. Could you also address the [pep8 issue above](https://github.com/dipy/dipy/pull/2488#issuecomment-992232863)?

See below for some comments.

skoudoro commented:
`test_sfm_background() `  is failing. more information [here](https://dev.azure.com/dipy/dipy/_build/results?buildId=1384&view=logs&j=2a814f24-3709-5a1b-1029-04051cd793c9&t=c6fe50ad-1d7b-5c9d-ed7d-e9f17140505a&l=5149).  
You can also run the tests locally to see the error. `pytest -svv dipy/reconst/tests/test_sfm.py`

dPys commented:
> `test_sfm_background() ` is failing. more information [here](https://dev.azure.com/dipy/dipy/_build/results?buildId=1384&view=logs&j=2a814f24-3709-5a1b-1029-04051cd793c9&t=c6fe50ad-1d7b-5c9d-ed7d-e9f17140505a&l=5149). You can also run the tests locally to see the error. `pytest -svv dipy/reconst/tests/test_sfm.py`

This is now addressed in `1d108b01a`. All tests should now be passing.

Formatting of docstring has also been cleaned up in a few places, and I've added a parallel_backend argument to `fit` so that we can tinker with backends other than `multiprocessing` (e.g. `loky` or `threading`) down the road if needed. Assuming @Garyfallidis is on-board for joblib and nobody catches anything else, then I feel confident this PR is ready to merge.

skoudoro commented:
we still have some tests failing @dPys. see [here](https://dev.azure.com/dipy/dipy/_build/results?buildId=1385&view=logs&j=a846d25a-e32c-5640-1b53-e815fab94407&t=2f8d848e-c100-586c-cef4-f5504e3b245e&l=4016)
```python
>       if not mask:
E       ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

dPys commented:
Should be good now. I changed line 506 by mistake when I shouldn't have! 🤞

skoudoro commented:
Hi @dPys,

Happy New Year! I hope you are doing well. Let's finish this PR!  I am afraid that your tests is no enough deterministic. 
All the tests are almost passing apart from this One matrix which is still failing.

Could you make your test more robust? Thanks!

see the error below:

```python
================================== FAILURES ===================================
__________________________________ test_sfm ___________________________________

    @needs_sklearn
    def test_sfm():
        fdata, fbvals, fbvecs = dpd.get_fnames()
        data = load_nifti_data(fdata)
        gtab = grad.gradient_table(fbvals, fbvecs)
        for n_procs in [1, 2]:
            for iso in [sfm.ExponentialIsotropicModel, None]:
                sfmodel = sfm.SparseFascicleModel(gtab, isotropic=iso)
                sffit1 = sfmodel.fit(data[0, 0, 0], num_processes=n_procs)
                sphere = dpd.get_sphere()
                odf1 = sffit1.odf(sphere)
                pred1 = sffit1.predict(gtab)
                mask = np.ones(data.shape[:-1])
                sffit2 = sfmodel.fit(data, mask, num_processes=n_procs)
                pred2 = sffit2.predict(gtab)
                odf2 = sffit2.odf(sphere)
                sffit3 = sfmodel.fit(data, num_processes=n_procs)
                pred3 = sffit3.predict(gtab)
                odf3 = sffit3.odf(sphere)
>               npt.assert_almost_equal(pred3, pred2, decimal=2)
E               AssertionError: 
E               Arrays are not almost equal to 2 decimals
E               
E               Mismatched elements: 4727 / 65000 (7.27%)
E               Max absolute difference: 237.76250004954414
E               Max relative difference: 28.828423844986176
E                x: array([[[[  8.90e+01,   4.22e+01,   4.15e+01, ...,   4.04e+01,   4.38e+01,
E                           3.65e+01],
E                        [  9.80e+01,   2.85e+01,   4.27e+01, ...,   3.93e+01,   3.33e+01,...
E                y: array([[[[  8.90e+01,   4.22e+01,   4.15e+01, ...,   4.04e+01,   4.38e+01,
E                           3.65e+01],
E                        [  9.80e+01,   2.85e+01,   4.27e+01, ...,   3.93e+01,   3.33e+01,...

dipy\reconst\tests\test_sfm.py:51: AssertionError
```

dPys commented:
Hi @skoudoro . Thanks for the heads up, though not quite sure why this was failing since it was passing locally:

```
collecting ... 

 dipy/reconst/tests/test_sfm.py::test_design_matrix ✓                                                                                                                     14% █▌        
 dipy/reconst/tests/test_sfm.py::test_sfm ✓                                                                                                                               29% ██▉       
 dipy/reconst/tests/test_sfm.py::test_predict ✓                                                                                                                           43% ████▍     
 dipy/reconst/tests/test_sfm.py::test_sfm_background ✓                                                                                                                    57% █████▊    
 dipy/reconst/tests/test_sfm.py::test_sfm_stick ✓                                                                                                                         71% ███████▎  
 dipy/reconst/tests/test_sfm.py::test_sfm_sklearnlinearsolver ✓                                                                                                           86% ████████▋ 
 dipy/reconst/tests/test_sfm.py::test_exponential_iso ✓                                                                                                                  100% ██████████
=================================================================================== warnings summary ===================================================================================
/usr/local/anaconda3/lib/python3.7/site-packages/socks.py:58
  /usr/local/anaconda3/lib/python3.7/site-packages/socks.py:58: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
    from collections import Callable

/usr/local/anaconda3/lib/python3.7/site-packages/pandas/compat/_optional.py:138
  /usr/local/anaconda3/lib/python3.7/site-packages/pandas/compat/_optional.py:138: UserWarning: Pandas requires version '2.7.0' or newer of 'numexpr' (version '2.6.9' currently installed).
    warnings.warn(msg, UserWarning)

-- Docs: https://docs.pytest.org/en/latest/warnings.html

Results (97.19s):
       7 passed
 ```
 
 I just pushed it again so let's see what happens?

dPys commented:
Closing (t.b.c with PR #2510 )
